File: frontend/src/types/results24.ts
/** Domain types for results / ResultRecord24 */
export type ResultRecord24Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface ResultRecord24 {
  id: string;
  externalCode: string;
  displayName: string;
  status: ResultRecord24Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface ResultRecord24CreateRequest {
  externalCode: string;
  displayName: string;
  status: ResultRecord24Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface ResultRecord24ListResponse {
  items: ResultRecord24[];
  total?: number;
}

export interface ResultRecord24StatsResponse {
  activeCount: number;
}

export function isResultRecord24Active(rec: ResultRecord24): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatResultRecord24Label(rec: ResultRecord24): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareResultRecord24ByPriority(a: ResultRecord24, b: ResultRecord24): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/results_24.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// ResultRecord24Service encapsulates business rules for ResultRecord24.
type ResultRecord24Service struct {
	repo *repository.ResultRecord24Repository
}

// NewResultRecord24Service wires the service.
func NewResultRecord24Service(repo *repository.ResultRecord24Repository) *ResultRecord24Service {
	return &ResultRecord24Service{repo: repo}
}

// Create validates and persists a new ResultRecord24.
func (s *ResultRecord24Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.ResultRecord24, error) {
	now := time.Now().UTC()
	rec := &models.ResultRecord24{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a ResultRecord24 by id.
func (s *ResultRecord24Service) Get(ctx context.Context, id string) (*models.ResultRecord24, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped ResultRecord24 pages.
func (s *ResultRecord24Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.ResultRecord24, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing ResultRecord24.
func (s *ResultRecord24Service) Update(ctx context.Context, actor string, rec *models.ResultRecord24) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a ResultRecord24.
func (s *ResultRecord24Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds ResultRecord24 by external code prefix.
func (s *ResultRecord24Service) Search(ctx context.Context, facilityID, prefix string) ([]models.ResultRecord24, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *ResultRecord24Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
